.config/crafted-emacs/config.org
#+title: Crafted Emacs Configuration
#+PROPERTY: header-args:emacs-lisp :tangle ./config.el :mkdirp yes

This configuration is built on top of [[https://github.com/SystemCrafters/crafted-emacs][Crafted Emacs]].

See the [[https://orgmode.org/manual/Extracting-Source-Code.html][Org Manual]] for details and alternatives on tangling code blocks.

#+begin_src emacs-lisp
  ;;; config.el -- Generated Crafted Emacs user configuration -*- lexical-binding: t; -*-
  ;; This file is generated from config.org. If you want to edit the
  ;; configuration, DO NOT edit config.el, edit config.org, instead.

  ;; Tangle the code blocks to config.el on save.
  (defun org-babel-tangle-config ()
    (when (string-equal (buffer-file-name)
                        (expand-file-name "config.org" crafted-config-path))
      ;; Dynamic scoping to the rescue
      (let ((org-confirm-babel-evaluate nil))
        (org-babel-tangle))))

  (add-hook 'org-mode-hook
            (lambda ()
              (add-hook 'after-save-hook #'org-babel-tangle-config)))
#+end_src

In this Org file we can set any Emacs configuration variable, face
attributes, themes, etc, as we normally would.
See the README.org file in [[https://github.com/SystemCrafters/crafted-emacs][Crafted Emacs]] repository for additional information.

* Crafted Modules

At the moment, Crafted Emacs offers the following modules. Comment out
everything you don't want to use.
At the very least, you should decide whether or not you want to use [[https://github.com/emacs-evil/evil][evil-mode]],
as it will greatly change how you interact with Emacs. So, if you prefer
Vim-style keybindings over vanilla Emacs keybindings remove the comment
in the line about =crafted-evil= below.
#+begin_src emacs-lisp
  (require 'crafted-defaults)    ; Sensible default settings for Emacs
  ;(require 'crafted-updates)     ; Tools to upgrade Crafted Emacs
  (require 'crafted-completion)  ; selection framework based on `vertico`
  (require 'crafted-ui)          ; Better UI experience (modeline etc.)
  ;(require 'crafted-windows)     ; Window management configuration
  (require 'crafted-editing)     ; Whitspace trimming, auto parens etc.
  ;(require 'crafted-evil)        ; An `evil-mode` configuration
  (require 'crafted-org)         ; org-appear, clickable hyperlinks etc.
  ;(require 'crafted-project)     ; built-in alternative to projectile
  ;(require 'crafted-speedbar)    ; built-in file-tree
  ;(require 'crafted-screencast)  ; show current command and binding in modeline
  ;(require 'crafted-compile)     ; automatically compile some emacs lisp files
#+end_src

* Font and theme settings

** Font settings

Set the default face. The default face is the basis for most other
faces used in Emacs. A "face" is a configuration including font,
font size, foreground and background colors and other attributes.
The fixed-pitch and fixed-pitch-serif faces are monospace faces
generally used as the default face for code. The variable-pitch
face is used when `variable-pitch-mode' is turned on, generally
whenever a non-monospace face is preferred.

Archcraft pre-installed:
"JetBrains Mono", "FiraCode Nerd Font", "DejaVu Sans Mono", "Source Code Pro".

#+begin_src emacs-lisp
  (add-hook 'emacs-startup-hook
            (lambda ()
              (custom-set-faces
               `(default ((t (:font "Input Mono Light 18" :height 130))))
               `(fixed-pitch ((t (:inherit (default)))))
               `(fixed-pitch-serif ((t (:inherit (default)))))
               `(variable-pitch ((t (:font "Liberation Sans 18")))))))
#+end_src

** Theme settings

Themes are color customization packages which coordinate the
various colors, and in some cases, font-sizes for various aspects
of text editing within Emacs, toolbars, tabbars and
modeline. Several themes are built-in to Emacs, by default,
Crafted Emacs uses the `deeper-blue' theme. Here is an example of
loading a different theme from the venerable Doom Emacs project.


#+begin_src emacs-lisp
  (crafted-package-install-package 'doom-themes)

  (setq doom-theme '(color-theme-solarized-light))

  (defun synchronize-theme ()
    (let* ((light-theme 'doom-ayu-light)
           (dark-theme 'doom-palenight)
           (start-time-light-theme 6)
           (end-time-light-theme 18)
           (hour (string-to-number (substring (current-time-string) 11 13)))
           (next-theme (if (member hour (number-sequence start-time-light-theme end-time-light-theme))
                           light-theme dark-theme)))
      (when (not (equal doom-theme next-theme))
        (disable-theme doom-theme)
        (setq doom-theme next-theme)
        (load-theme next-theme))))

  (run-with-timer 0 900 'synchronize-theme)
#+end_src

https://www.reddit.com/r/emacs/comments/jqeskb/how_can_i_set_themes_depending_on_time_of_the_day/
https://stackoverflow.com/questions/14760567/emacs-auto-load-color-theme-by-time

* Keyboard Shortcuts

#+begin_src emacs-lisp
  (crafted-ensure-package 'ergoemacs-mode)
  (require 'ergoemacs-mode)
  (ergoemacs-mode t)
#+end_src

This can replace =which-key=.

Usage ex.:
1. press =Ctrl+c=
2. press Ctrl+h
3. filter by ='roam'=
4. navigate with =M-k= or =M-i= to desired command and press =Enter=
   1. or, type at sign =@= and press the rest of the key chord, eg. =n f=.

#+begin_src emacs-lisp
  (setq prefix-help-command #'embark-prefix-help-command)
#+end_src

Ref:
[[https://youtu.be/qk2Is_sC8Lk?t=5502][System Crafters Live! - The Many Uses of Embark]]
He set's it under =:init= keyword with =use-package=.
Should it be set before loading Embark?
How would we do it?
Crafted Emacs loads it.

* Org-mode
This enables us to type =<s + TAB= to expand to a =#+begin_src ... #+end_src= block.
https://github.com/syl20bnr/spacemacs/issues/11798#issuecomment-454941024

#+begin_src emacs-lisp
  (when (version<= "9.2" (org-version))
    (require 'org-tempo))
#+end_src

#+begin_src emacs-lisp
  (setq org-startup-folded t)
  ;; Soft-wrapping
  (global-visual-line-mode)
  (diminish 'visual-line-mode)
#+end_src

I like indenting text to the same level has their headings:

#+begin_src emacs-lisp
  (setq org-adapt-indentation t)
#+end_src

** Org-roam

   #+begin_src emacs-lisp
     (crafted-ensure-package 'org-roam)
     (setq-default org-roam-directory (file-truename "~/Notes"))
     (setq org-roam-dailies-directory "Journal/")

     (setq org-roam-dailies-capture-templates
           '(("d" "default" entry
              "\n* %<%H:%M> %?"
              :if-new (file+head "%<%Y/%m-%d>.org"
                                 "#+title: %<%a, %b %d %Y>\n"))))

     (require 'org-roam)
     (with-eval-after-load 'org-roam
       (org-roam-db-autosync-mode))

     (define-key org-mode-map (kbd "C-c n i") 'org-roam-node-insert)
     (global-set-key (kbd "C-c n c") 'org-roam-capture)
     (global-set-key (kbd "C-c n f") 'org-roam-node-find)
     (global-set-key (kbd "C-c n j") 'org-roam-dailies-capture-today)
   #+end_src

   =org-roam-dailies= refs:
   - https://systemcrafters.net/build-a-second-brain-in-emacs/keep-a-journal/

   Why isn't ="\n"= not working in =org-roam-dailies-capture-templates=?

   #+begin_src bash
     ln -s ~/Mega/Data1/Org.d/Roam2/ ~/Notes
   #+end_src

   After setting up  =org-roam-dailies=, I got the following error:
   "Could not read ‘org-id-values’ from /home/iqb/.config/crafted-emacs/var/org-id-locations, setting it to nil"
   
   To fix, I had to run:

   #+begin_src emacs-lisp :tangle no
     (org-id-update-id-locations (directory-files-recursively org-roam-directory ".org$\\|.org.gpg$"))
   #+end_src
   Ref: https://github.com/org-roam/org-roam/issues/1700#issuecomment-889116709

   https://github.com/daviwil/dotfiles/blob/0e034ebb688633082a3b0c201b0bc834a9821091/Emacs.org#org-roam
   https://jethrokuan.github.io/org-roam-guide/
   https://lucidmanager.org/productivity/taking-notes-with-emacs-org-mode-and-org-roam/
   
* Custom.el
By default, Crafted Emacs keeps your config file clean. All the customization
settings that Emacs normally automatically adds to your config.el go into
the file =custom.el= instead. If you don't want this, set the respective
variable to =nil=:
#+begin_src emacs-lisp
  ;; To not load `custom.el' after `config.el', uncomment this line.
  ;; (setq crafted-load-custom-file nil)
#+end_src

* Scrolling

** Cursor Guiding

   "A light that follows your cursor around so you don't lose it!"

   #+begin_src emacs-lisp
     (crafted-ensure-package 'beacon)
     (beacon-mode 1)
     (diminish 'beacon-mode " ⓑ")
   #+end_src
   Diminishing alternatives: Ⓑ ⓑ ⒝ ḇ Ḇ ḅ Ḅ ɓ Ƀ ƅ Ƅ ƃ Ƃ Ɓ ß ƀ

   *Guix Packages*

   #+begin_src scheme :noweb-ref packages :nowebsep ""
     ; "emacs-beacon"
   #+end_src

* Tangling to early-config.el
If you need to make settings to =early-config.el=, you can do that from here, too.
Just begin the source code block with:
#+begin_src org
  ,#+begin_src emacs-lisp :tangle ./early-config.el
#+end_src

* TODOs

** TODO Set up *Guix Packages* file tangling
   https://github.com/daviwil/dotfiles/blob/0e034ebb688633082a3b0c201b0bc834a9821091/Emacs.org

* Versioning

  #+begin_src emacs-lisp :tangle no
    ;;(add-to-list 'embark-vc-file-map '("v" vc-next-act))
    (eval-after-load 'embark
      (define-key embark-vc-file-map "s" 'magit-stage-file))
  #+end_src

* Windows

#+begin_src emacs-lisp  
  (if (daemonp)
      (add-to-list 'default-frame-alist '(fullscreen . fullboth))
    (toggle-frame-fullscreen))
#+end_src


.config/crafted-emacs-v1/config.el
;;; config.el -- Generated Crafted Emacs user configuration -*- lexical-binding: t; -*-
;; This file is generated from config.org. If you want to edit the
;; configuration, DO NOT edit config.el, edit config.org, instead.

;; Tangle the code blocks to config.el on save.
(defun org-babel-tangle-config ()
  (when (string-equal (buffer-file-name)
                      (expand-file-name "config.org" crafted-config-path))
    ;; Dynamic scoping to the rescue
    (let ((org-confirm-babel-evaluate nil))
      (org-babel-tangle))))

(add-hook 'org-mode-hook
          (lambda ()
            (add-hook 'after-save-hook #'org-babel-tangle-config)))

(require 'crafted-defaults)    ; Sensible default settings for Emacs
;(require 'crafted-updates)     ; Tools to upgrade Crafted Emacs
(require 'crafted-completion)  ; selection framework based on `vertico`
(require 'crafted-ui)          ; Better UI experience (modeline etc.)
;(require 'crafted-windows)     ; Window management configuration
(require 'crafted-editing)     ; Whitspace trimming, auto parens etc.
;(require 'crafted-evil)        ; An `evil-mode` configuration
(require 'crafted-org)         ; org-appear, clickable hyperlinks etc.
;(require 'crafted-project)     ; built-in alternative to projectile
;(require 'crafted-speedbar)    ; built-in file-tree
;(require 'crafted-screencast)  ; show current command and binding in modeline
;(require 'crafted-compile)     ; automatically compile some emacs lisp files

(add-hook 'emacs-startup-hook
          (lambda ()
            (custom-set-faces
             `(default ((t (:font "Input Mono Light 18" :height 130))))
             `(fixed-pitch ((t (:inherit (default)))))
             `(fixed-pitch-serif ((t (:inherit (default)))))
             `(variable-pitch ((t (:font "Liberation Sans 18")))))))

(crafted-package-install-package 'doom-themes)

(setq doom-theme '(color-theme-solarized-light))

(defun synchronize-theme ()
  (let* ((light-theme 'doom-ayu-light)
         (dark-theme 'doom-palenight)
         (start-time-light-theme 6)
         (end-time-light-theme 18)
         (hour (string-to-number (substring (current-time-string) 11 13)))
         (next-theme (if (member hour (number-sequence start-time-light-theme end-time-light-theme))
                         light-theme dark-theme)))
    (when (not (equal doom-theme next-theme))
      (disable-theme doom-theme)
      (setq doom-theme next-theme)
      (load-theme next-theme))))

(run-with-timer 0 900 'synchronize-theme)

(crafted-ensure-package 'ergoemacs-mode)
(require 'ergoemacs-mode)
(ergoemacs-mode t)

(setq prefix-help-command #'embark-prefix-help-command)

(when (version<= "9.2" (org-version))
  (require 'org-tempo))

(setq org-startup-folded t)
;; Soft-wrapping
(global-visual-line-mode)
(diminish 'visual-line-mode)

(setq org-adapt-indentation 'headline-data)

(crafted-ensure-package 'org-roam)
(setq-default org-roam-directory (file-truename "~/Notes"))
(setq org-roam-dailies-directory "Journal/")

(setq org-roam-dailies-capture-templates
      '(("d" "default" entry
         "\n* %<%H:%M> %?"
         :if-new (file+head "%<%Y/%m-%d>.org"
                            "#+title: %<%a, %b %d %Y>\n"))))

(require 'org-roam)
(with-eval-after-load 'org-roam
  (org-roam-db-autosync-mode))

(define-key org-mode-map (kbd "C-c n i") 'org-roam-node-insert)
(global-set-key (kbd "C-c n c") 'org-roam-capture)
(global-set-key (kbd "C-c n f") 'org-roam-node-find)
(global-set-key (kbd "C-c n j") 'org-roam-dailies-capture-today)

;; To not load `custom.el' after `config.el', uncomment this line.
;; (setq crafted-load-custom-file nil)

(crafted-ensure-package 'beacon)
(beacon-mode 1)
(diminish 'beacon-mode " ⓑ")

(if (daemonp)
    (add-to-list 'default-frame-alist '(fullscreen . fullboth))
  (toggle-frame-fullscreen))
